Fix no-sm2

mattcaswell opened the pull request:
Fixes #6525

richsalz commented on test/evp_test.c:
```diff
@@ -2516,13 +2499,23 @@ static int parse(EVP_TEST *t)
+            TEST_info("Line %d: missing blank line\n", t->s.curr);
+
+        /* Hack to detect SM2 keys */
+        if(strstr(pp->value, "SM2")) {
```
Add explicit NULL test.

tmshort commented on test/evp_test.c:
```diff
@@ -2531,17 +2524,11 @@ static int parse(EVP_TEST *t)
-        if(strstr(key->name, "SM2"))
-            EVP_PKEY_set_alias_type(pkey, EVP_PKEY_SM2);
-
         key->key = pkey;
```
When pkey = NULL, tests fail because the key can be found, and it's NULL, so operations fail. At least, that's what I saw. I will download this to see if it fixes the problem I saw.

tmshort commented:
I'm still seeing a failure with `no-sm2`with this patch:

```
$ ./config no-sm2
...
$ make -sj
...
$ make test V=1 TESTS=test_evp
...
        # INFO:  @ test/testutil/stanza.c:121
        # Starting "RFC7919 DH tests" tests at line 17816
        # INFO:  @ test/testutil/stanza.c:121
        # Starting "SM2 tests" tests at line 18373
../../util/shlib_wrap.sh ../../test/evp_test ../../test/recipes/30-test_evp_data/evppkey.txt => 139
not ok 7 - running evp_test evppkey.txt

#   Failed test 'running evp_test evppkey.txt'
#   at ../test/recipes/30-test_evp.t line 23.
# Looks like you failed 1 test of 7.
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/7 subtests 
```

mattcaswell commented on test/evp_test.c:
```diff
@@ -2516,13 +2499,23 @@ static int parse(EVP_TEST *t)
+
+        /* Hack to detect SM2 keys */
+        if(strstr(pp->value, "SM2")) {
+#ifdef OPENSSL_NO_SM2
```
This SM2 specific code is there in your patch too? I just moved it.

Or are you referring specifically to the `OPENSSL_NO_SM2` guard? If so, I think it is unavoidable. Before it was still there - but was inside the key_disabled() function. I've just removed that function and moved the guard here instead. Since this is in SM2 specific code anyway, having an `OPENSSL_NO_SM2` guard doesn't seem unreasonable to me.

mattcaswell commented:
> I'm still seeing a failure with no-sm2with this patch:

Hmm...I don't see that??

mattcaswell commented:
Just to further note that the existing `key_disabled` test is wrong now that SM2 has been moved into its own EVP_PKEY_METHOD - so it needs to be fixed (or just deleted as in this patch).

tmshort commented:
I pulled your changes straight onto master w/o my stuff. 
The key creation returns NULL, and klist is unconditionally set; I had a change that modified that, but undid it. 
I will re-verify the patch and the behavior later...

mattcaswell commented:
I fixed @richsalz's style nit, and also simplified this patch a little. On further investigation it turns out that if the pkey ends up being NULL its still ok to add it to the list of keys. Any tests that use that key get skipped later. The problem was the unconditional deref of that pkey in the `EVP_PKEY_set_alias_type()` call.

@richsalz can you reconfirm?

@tmshort can you check again whether you are still seeing a problem with this patch?

mattcaswell commented:
Pushed. Thanks all.
